fix(api): resolve TypeScript errors in webviewWindow.ts

KushalMeghani1644 opened the pull request:
## Fix TypeScript errors in `webviewWindow.ts`

### Problem
The `webviewWindow.ts` file contained several TypeScript issues:
- Conflicts when extending both `WebviewOptions` and `WindowOptions` due to overlapping properties.
- The `skip` option was being used internally but was missing from the type definitions.
- The `parent` property was not properly typed on the internal options interface.

### Solution
- Introduced `InternalWebviewOptions` to extend only non-conflicting properties from `WebviewOptions`.
- Added `skip` and `parent` properties explicitly to the internal interface.
- Updated the constructor parameter type to use `InternalWebviewOptions`.
- Replaced `@ts-expect-error` usages with proper typing and safe type assertions.

### Changes
- Maintains existing functionality and API compatibility.
- No breaking changes to the public API.

---

_AI disclosure: This PR description was polished with the help of AI. The code, commit message, and logic were written without AI assistance._

KushalMeghani1644 committed:
fix(api): resolve TypeScript errors in webviewWindow.ts

Fix interface conflicts between WebviewOptions and WindowOptions by properly
handling type intersections and adding missing properties to internal types.

lucasfernog commented on packages/api/src/webviewWindow.ts:
```diff
@@ -53,43 +58,24 @@ class WebviewWindow {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   listeners: Record<string, Array<EventCallback<any>>>
 
-  /**
```
can you readd the documentation code?

KushalMeghani1644 replied:
Indeed, why not! if the documentation code is needed then yes, I would love to add it back!

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

KushalMeghani1644 committed:
Added back the documentation.

KushalMeghani1644 committed:
Merge branch 'webviewWindowFIX' of github.com:KushalMeghani1644/tauri into webviewWindowFIX

Legend-Master commented on packages/api/src/webviewWindow.ts:
```diff
@@ -109,30 +113,27 @@ class WebviewWindow {
     if (webview) {
-      // @ts-expect-error `skip` is not defined in the public API but it is handled by the constructor
-      return new WebviewWindow(webview.label, { skip: true })
+      return new WebviewWindow(webview.label, { skip: true } as any)
```
To be honest, using as any instead of expect error doesn't feel like an improvement to me

KushalMeghani1644 replied:
Thanks for pointing that out, and you are right! `as any` weakens type safety, I'll revert that to @ts-expect-error. Thanks for putting time into my PR.

KushalMeghani1644 committed:
Removed  to strengthen type-safety

KushalMeghani1644 commented:
Sorry for the weird looking commit message, I accidentally included a backtick. xD

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -75,21 +79,21 @@ class WebviewWindow {
     label: WebviewLabel,
     options: Omit<WebviewOptions, 'x' | 'y' | 'width' | 'height'>
-      & WindowOptions = {}
+      & WindowOptions & { skip?: boolean } = {}
```
This will be visible in the public API, which is why we didn't do it in the first place, as this concerns internal usage only.

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -15,14 +15,20 @@ import type { EventName, EventCallback, UnlistenFn } from './event'
+interface InternalWebviewOptions
+  extends Omit<WebviewOptions, 'x' | 'y' | 'width' | 'height'> {
+  skip?: boolean
+  parent?: Window | WebviewWindow | string
```
Why redefine parent here, it should exist and inherited from `WebviewOptions`

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -122,17 +126,14 @@ class WebviewWindow {
   /**
    * Get an instance of `Webview` for the current webview.
    */
+
```
why this newline here?

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -122,17 +126,14 @@ class WebviewWindow {
     return getCurrentWebviewWindow()
   }
 
-  /**
-   * Gets a list of instances of `Webview` for all available webviews.
-   */
```
why remove this doc comment?

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -122,17 +126,14 @@ class WebviewWindow {
   static async getAll(): Promise<WebviewWindow[]> {
     return getAllWebviewWindows()
   }
-
```
why remove this newline? run `pnpm format` in repo root to fix these formatting issue

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -187,46 +189,57 @@ class WebviewWindow {
    * @returns A promise resolving to a function to unlisten to the event.
    * Note that removing the listener is required if your listener goes out of scope e.g. the component is unmounted.
    */
+
```
ditto, run `pnpm format`

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -187,46 +189,57 @@ class WebviewWindow {
   /**
    * Set the window and webview background color.
    *
-   * #### Platform-specific:
-   *
-   * - **Android / iOS:** Unsupported for the window layer.
-   * - **macOS / iOS**: Not implemented for the webview layer.
-   * - **Windows**:
-   *   - alpha channel is ignored for the window layer.
-   *   - On Windows 7, alpha channel is ignored for the webview layer.
-   *   - On Windows 8 and newer, if alpha channel is not `0`, it will be ignored.
-   *
-   * @returns A promise indicating the success or failure of the operation.
-   *
```
why is this documentation removed?

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -187,46 +189,57 @@ class WebviewWindow {
+    return once(
+      event,
+      handler,
+      /**
+       * Set the window and webview background color.
+       *
+       * #### Platform-specific:
+       *
+       * - **Android / iOS:** Unsupported for the window layer.
+       * - **macOS / iOS**: Not implemented for the webview layer.
+       * - **Windows**:
+       *   - alpha channel is ignored for the window layer.
+       *   - On Windows 7, alpha channel is ignored for the webview layer.
+       *   - On Windows 8 and newer, if alpha channel is not `0`, it will be ignored.
+       *
+       * @returns A promise indicating the success or failure of the operation.
+       *
+       * @since 2.1.0
+       */ {
```
why add documentation here? it serves nothing

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -187,46 +189,57 @@ class WebviewWindow {
+      await invoke('plugin:window|set_background_color', { color })
+      await invoke('plugin:webview|set_webview_background_color', { color })
+    } catch (err) {
+      throw new Error(`Failed to set background color: ${String(err)}`)
```
Pass the error as cause instead
```suggestion
      throw new Error("Failed to set background color", { cause: err })
```

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -187,46 +189,57 @@ class WebviewWindow {
   }
 }
 
-// Order matters, we use window APIs by default
```
revert this comment, it is vital

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -75,21 +79,21 @@ class WebviewWindow {
     this.label = label
-    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
     this.listeners = Object.create(null)
 
-    // @ts-expect-error `skip` is not a public API so it is not defined in WebviewOptions
-    if (!options?.skip) {
+    const internalOptions = options as InternalWebviewOptions
+
+    if (!internalOptions?.skip) {
```
instead of defining a new `InternalWebviewOptions`, we can just do 
```ts
if (options && "skip" in options && options.skip) {
  // ...
}
```

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

KushalMeghani1644 committed:
fix(api): address webviewWindow.ts review comments

- Hide skip property from public API
- Remove redundant parent property from InternalWebviewOptions
- Restore getAllWebviewWindows doc comment
- Remove misplaced setBackgroundColor documentation
- Use error cause instead of string interpolation

KushalMeghani1644 committed:
Merge branch 'webviewWindowFIX' of github.com:KushalMeghani1644/tauri into webviewWindowFIX

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

amrbashir commented on packages/api/package.json:
```diff
@@ -55,6 +55,7 @@
     "eslint-plugin-security": "3.0.1",
     "fast-glob": "3.3.3",
     "globals": "^16.2.0",
+    "prettier": "^3.5.3",
```
No need to add prettier as dependency here, it is already included in the workspace package.json

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -15,30 +15,32 @@ import type { EventName, EventCallback, UnlistenFn } from './event'
+  extends Omit<WebviewOptions, 'x' | 'y' | 'width' | 'height'> {
+  skip?: boolean
+}
+
```
This is no longer needed, let's remove it

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -15,30 +15,32 @@ import type { EventName, EventCallback, UnlistenFn } from './event'
-          // @ts-expect-error `skip` is not defined in the public API but it is handled by the constructor
           skip: true
-        })
+        } as any)
```
Honestly I would still prefer to use `@ts-expext-error` over `as any` as with the former we have a comment explaining why we bypassing typescript

KushalMeghani1644 committed:
refactor: remove InternalWebviewOptions interface in webviewWindow.ts

- Remove InternalWebviewOptions interface to simplify code
- Replace with inline type checking using 'skip' in options
- Add skip?: boolean to constructor parameter type
- Remove unnecessary type assertions and @ts-expect-error comments

KushalMeghani1644 committed:
Merge branch 'webviewWindowFIX' of github.com:KushalMeghani1644/tauri into webviewWindowFIX

KushalMeghani1644 commented:
Hey! @amrbashir I have fixed the WebViewWindow.ts and have made all the changes as requested, if there are more improvements remaining, I'd be happy to make them :)

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -206,27 +197,20 @@ class WebviewWindow {
+    } catch (err) {
+      const error = new Error('Failed to set background color')
+      ;(error as any).cause = err
+      throw error
```
why are you setting the `cause` field like this, it should be 
```suggestion
      throw new Error('Failed to set background color', { cause: err })
```

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -22,21 +22,17 @@ import type { Color, DragDropEvent } from './webview'
- * Gets a list of instances of `Webview` for all available webview windows.
- *
- * @since 2.0.0
+ * Gets a list of instances of `Webview` for all available webviews.
```
Please revert this

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -206,27 +197,20 @@ class WebviewWindow {
   /**
    * Set the window and webview background color.
    *
-   * #### Platform-specific:
-   *
-   * - **Android / iOS:** Unsupported for the window layer.
-   * - **macOS / iOS**: Not implemented for the webview layer.
-   * - **Windows**:
-   *   - alpha channel is ignored for the window layer.
-   *   - On Windows 7, alpha channel is ignored for the webview layer.
-   *   - On Windows 8 and newer, if alpha channel is not `0`, it will be ignored.
-   *
-   * @returns A promise indicating the success or failure of the operation.
-   *
```
Please revert this

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -206,27 +197,20 @@ class WebviewWindow {
   }
 }
 
-// Order matters, we use window APIs by default
```
Please revert this comment, it is vital

amrbashir commented on packages/api/src/webviewWindow.ts:
```diff
@@ -126,13 +119,9 @@ class WebviewWindow {
     return getCurrentWebviewWindow()
   }
 
-  /**
-   * Gets a list of instances of `Webview` for all available webviews.
-   */
```
Please revert this

amrbashir commented:
Please run `pnpm format` to fix all the formatting issue and there are multiple places where you removed documentation for no reason, please revert those.

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

KushalMeghani1644 committed:
Update packages/api/src/webviewWindow.ts

Co-authored-by: Amr Bashir <github@amrbashir.me>

KushalMeghani1644 committed:
Update webviewWindow.ts, added comments.

KushalMeghani1644 commented:
Hey @amrbashir I have reverted all the things as you have requested and committed your suggestions :) I hope the changes are fine and if there are more improvements, I'd be happy to make them :)

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

KushalMeghani1644 committed:
Merge branch 'dev' into webviewWindowFIX

Legend-Master commented:
Sorry but I am closing this for now because of the following reasons:

- Unrelated changes that doesn't make sense (e.g. removal of documentations)
- Applying `{ skip?: boolean }` in the constructor exposes that to the public API which amr has pointed out already, the previous solution simply used `as any` which didn't help improving anything either

Feel free to open a new PR if that solves the above problems
